### PR TITLE
patch out execinfo/backtrace outside gnu libc on linux

### PR DESCRIPTION
**Pull request by selfisekai**
fixes one of build errors for musl targets

*List which issues are fixed by this PR. You must list at least one issue.*
https://github.com/flutter/flutter/issues/124285

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`) - n/a?
- [x] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by cbracken — CHANGES_REQUESTED** [review]
The right solution for those using musl is to build with `backtrace_stub.cc` instead of `backtrace.cc`. See: https://github.com/flutter/engine/blob/92c75660eb192c9ace0adae284171bf69c884c51/fml/BUILD.gn#L100-L105

I'm not aware of any build variable we might have that helps spot which libc is in use. This sort of thing is typically declared in:
https://github.com/flutter/buildroot/blob/master/build/config/BUILDCONFIG.gn. 

If not, I suspect the best approach would be:
* Add a `--musl` flag for `tools/gn` that sets a gn variable like `has_musl`. Error out if anyone uses the option on anything but Linux builds.
* Tweak the expression above in BUILD.gn to use backstrace_stub.cc when that's set.

@chinmaygarde or @zanderso may have additional thoughts.

**Comment by selfisekai** [comment]
afaik the only gn-based project that acknowledges musl targets is dart, but that's only through `dart_sysroot == "alpine"`, and it still does not acknowledge musl builds on musl hosts

File: fml/backtrace.cc [hunk 1]
```diff
@@ -35,9 +35,10 @@ static std::string GetSymbolName(void* symbol) {
 static int Backtrace(void** symbols, int size) {
 #if FML_OS_WIN
   return CaptureStackBackTrace(0, size, symbols, NULL);
-#else
+#elif !defined(FML_OS_LINUX) || defined(__GLIBC__)
   return ::backtrace(symbols, size);
 #endif  // FML_OS_WIN
+  // do nothing on musl
```
Comment (chinmaygarde) [inline fml/backtrace.cc:41]: We can leave out the references to `musl` in the comments.

**Review by chinmaygarde — APPROVED** [review]

**Comment by cbracken** [comment]
Since we already have backtrace_stub.cc for these types of situation, I'm proposing https://github.com/flutter/engine/pull/40980 as an alternative, which avoids backtrace.cc being half-stubbed-out in some situations, and makes the build files a bit more readable in the process.

**Comment by cbracken** [comment]
Swapped out for https://github.com/flutter/engine/pull/40980. Thanks for identifying the issue and as noted in my comment on the other patch, if you could file an issue that provides an overview of the issues you're hitting compiling for Alpine, please go for it and CC me. I'd be glad to work with you to minimise the pain of porting.

**Pull request closed without merging**